### PR TITLE
CFC legibility in the console: per-cell labels cross-referenced from the space (CT-2100)

### PR DESCRIPTION
**Pull request by bfollington**
## The gap

Per-cell CFC labels exist today, and CT-2076 (E3) proved the whole taint story reconstructs offline. But the labels live in the **space DB**, and the run's artifact tree does not cross-reference them. So every reader working from artifacts alone — the console run reader (#6453) among them — draws a cell with no confidentiality atoms on it, under `max-enforcement` with `cfcFlowLabels: persist`, whatever the run was enforcing. The only labels the console could show sat on `bash` arguments, because `HarnessCfcInvocationContext` is recorded for sandbox operations only.

CT-2076's own framing: *the story is split across the artifact tree and the space DB, which don't cross-reference each other.* This is the cross-reference.

## What this does

- **`src/space-labels.ts`** — a read-only reader over the space SQLite, through `@commonfabric/state-inspector`'s lens. Resolves the run's space by name/DID/path, reads each cell's `cfc` labelMap, and normalizes atoms, `origin`, `observes`, and the `TransformedBy` provenance atom.
- **`src/contracts/cell-labels.ts`** + **`artifacts.ts`** + **`engine.ts`** — `cell-labels.json`, written beside the run. The CLI takes the snapshot when a run's loop ends; the console when a turn does, over the run and its `delegate_task` children.
- **`console/`** — the run reader carries the labels onto every cell it draws, and the map heads with what the run read.
- **`packages/patterns/cfc-input-cell-demo/`** — a checked-in seed (`secret` declared confidential, `city` plain) and a briefing pattern deriving from both, plus the runbook.

## Verified end to end, against a live space

A matched-build toolshed, the seed deployed with `cf piece new`, both cells handed to a real cf-harness run as `--input-cell`, `max-enforcement` with flow labels persisting. The run's `cell-labels.json`:

| cell | atoms | origin | producer |
|---|---|---|---|
| seed `secret` | `demo-secret` | `declared` | — |
| seed `city` | — | — | — |
| piece registry | `demo-secret` | `link` (+ `LinkReference`) | — |
| result `briefing` | `demo-secret` | `derived` | `briefingOf` |
| result `climate` | — | — | — |

That is CT-2076's reconstruction — entry labels, entry edges, the transform, and the negative space — collapsed into one artifact, and rendered in the console: the result chip reads `cell labelled derived` with `transformed by briefingOf`, the seed chips read labelled but carried, and `city` reads bare under a header that says the labels *were* read.

## Three things the design turns on

**Read vs unread.** A cell with no atoms under a snapshot that was taken is a cell the space holds no label for. The same cell on a host with no copy of the space is a cell nobody asked about. `status` keeps them apart, and the map header states the run-level fact so a bare cell reads as a finding rather than a hole.

**Derived vs carried.** CT-2076's second run flagged this (F4): a result object's label is the join of its fields, so a plain input reached through a labelled object carries a confidentiality atom without having been derived. *A renderer that colors by the confidentiality label alone would mislabel plain inputs as tainted.* The chip keys on `origin` and the `TransformedBy` atom, and gives a derived cell its own state.

**A reference is not a document.** Labels are stored per document; a run holds a reference per cell, so a piece's `secret` and `city` are two references into one label set. Lookup narrows to the path a reference walks. Caught by rendering the live run — without it, `city` wore `secret`'s atom, which is F4 one level down.

## Also closes, incidentally

CT-2076 gap 1 — *artifacts never name the space (no `did:key:` anywhere)*. The snapshot records the space it read, its DID, and the database file.

## Hand-offs

- **Thread C (#6471):** `CfcRefusalDetail` (`gate`/`inputs`/`attribution`) renders in the same two places this PR establishes — the cell chip's card and the map header. The label plumbing here is the shape to follow; no change to this PR is needed first.
- **Follow-up, not done here:** `console/src/steps-view.ts` calls `consoleStepArguments(step, this.handles)` without the label index, so an argument written as a whole link that the run holds *no handle for* shows no labels in the timeline pane. Every cell reached through a handle is covered, which is every cell in practice.
- A scoped reference (`/of:…@user`) names a scope kind, not a principal, while the store partitions by principal — so a per-user override's own labels are not addressable from a reference alone, and the read falls back to the base scope. Documented at the point it matters.

## Review

Codex review returned five blocking findings, all verified and all now fixed — four of which had the page state something it could not support:

- **Cross-space references read from the wrong store.** Content-addressed ids repeat across spaces, so resolving a foreign reference against the opened store answers with a different cell's labels. Reproduced by the reviewer with a seeded local collision. Now fail-closed: the opened store's DID is established from the file it was opened from (so an explicit `--space-db` proves its space too), and any reference or one-hop link whose space cannot be *proven* equal is recorded `unreadReason: "cross-space"` and excluded from the console index — nothing known, rather than read-and-unlabelled.
- **Wildcard and JSON Pointer paths were lost to string joining**, so an entry at `items/*` and a key holding a separator each narrowed to nothing — a labelled cell drawn **bare** under a snapshot that read it. Paths are segment arrays end to end now, refs parse through `parseLLMFriendlyLink`, and matching mirrors the runner's `cfcLabelPathPrefixMatches`.
- **A family claimed the root's status for every descendant.** A family whose runs did not all read the space is now read for none of them, with the split in its detail; a run that minted no handle is left out of the vote.
- **The terminal event outran the snapshot**, so the page's one re-read could observe `absent` and stay stale. That envelope's fan-out is now deferred until the snapshot settles, chained so stream order holds, and a failed snapshot still lets the event through.
- **The console loaded sqlite eagerly**, bypassing the engine's lazy import. Now imported at the point of use, after the zero-ref guard.

## Coverage

flush-left marker below accepts the residual rise.

The original rise was 148 lines. Pattern tests for the demo pair paid off all 42 of `packages/patterns` — a pattern file with no LCOV record has every tracked line counted uncovered, so testing them removed the group entirely — and tests for the reader, the run store and the console server took `packages/cf-harness` down to **+22**.

What remains is dominated by the Lit render bodies of `console/src/cell-chip.ts` (50) and `console/src/flow-view.ts` (18). cf-harness has no browser-test setup; `packages/ui` reaches its components through `deno-web-test`, and adding that lane to cf-harness's suite for two view components — with the browser launches it brings — is not proportionate to what it would buy. The pure logic behind both views is unit-tested directly (`cellLabelView`, `cellChipClasses`, `cellName`, `schemaSummary`).

ACCEPT_COVERAGE_DEBT: packages/cf-harness +22 lines

Refs CT-2100, CT-2076, CT-2066 (item 6).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/cf-harness/src/space-labels.ts [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+ */
+export const cellAddressOfRef = (ref: string): CellAddress | undefined => {
+  const trimmed = ref.trim();
+  try {
+    const link = parseLLMFriendlyLink(
+      trimmed.startsWith("/") ? trimmed : `/${trimmed}`,
+    );
+    return link.id === undefined
+      ? undefined
+      : { id: link.id, scope: link.scope ?? "space" };
```
Comment (bfollington) [inline packages/cf-harness/src/space-labels.ts]: 🔴 Blocking (verified): this drops the embedded `link.space`, so a direct cross-space handle is looked up by entity id in the DB opened for the run. The nested-link path has the same failure when `spaceDbPath` is used: discovery is skipped, `did` is `undefined`, and the `space === undefined` arm at line 202 accepts every explicitly cross-space link before reconstructing its id from the current DB. A focused repro seeded the same target id locally beneath a foreign-space link and the reader grafted that local `foreign-secret` label onto the foreign reference. Please retain the parsed space in `CellAddress`, establish the opened DB's DID for explicit paths, and fail closed for any explicitly spaced direct or nested link whose equality to the opened DID cannot be proven. Add direct and nested id-collision tests.

File: packages/cf-harness/console/cell-labels.ts [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    cellsRead: cells.length,
+    cellsLabelled: cells.filter((labels) => labels.entries.length > 0).length,
+  };
+};
+
+/** The entity a reference names, and the path inside it the reference walks. */
+const addressOf = (
+  ref: string,
+): { entityId: string; path: string[] } | undefined => {
+  const segments = ref.split("/").filter((segment) => segment !== "");
```
Comment (bfollington) [inline packages/cf-harness/console/cell-labels.ts]: 🔴 Blocking (verified): splitting the reference and flattening stored paths with `join("/")` loses the path semantics CFC actually stores. JSON Pointer keys containing `/` or `~` no longer match their encoded references, and runtime label maps use `*` segments whose ancestor coverage must match concrete children. On this head, an entry at `["items", "*"]` for `/<entity>/items/0`, and an entry at `["a/b"]` for `/<entity>/a~1b`, both narrow to an empty label set. That renders labelled cells bare under a `read` snapshot. Please keep paths as segment arrays, parse refs with the canonical LLM-friendly-link parser, and use the existing wildcard-aware `cfcLabelPathPrefixMatches` semantics before re-rooting segment-wise. Add wildcard and RFC 6901 cases alongside the current ancestor/descendant/sibling tests.

**Review by bfollington — COMMENTED** [review]
cf-review via Codex (GPT-5), on behalf of @bfollington

Scope: PR 6472 · scope-and-theory review with deep reads of the label reader, artifact contract, engine boundary, console indexing/rendering, and family/event integration; the six requested invariants were independently cross-checked.

🔴 Blocking: five verified findings are anchored inline. The durable space is opened read-only and link traversal is structurally limited to one hop, but cross-space identity is not bounded to that store. Ordinary ancestor/descendant/sibling narrowing works, but valid wildcard and JSON Pointer paths do not. The derived-vs-carried presentation correctly keys derived state on `origin` and exposes `TransformedBy`, independently of confidentiality atoms.

Checks: focused tests passed (4 files, 89 steps); touched-file format and lint checks passed; focused `deno check` passed. Negative probes reproduced the cross-space collision, wildcard/escaped-path loss, and eager console SQLite load; importing the engine without taking a snapshot succeeded without FFI permission.

Verdict: changes are required before merge.

File: packages/cf-harness/console/run-store.ts [hunk 3]
```diff
@@ -334,6 +371,7 @@ const readConsoleRunFamily = async (
     root: withNeighbours({ runId, steps: root.steps, handles: root.handles }),
     descendants: descendants.map(withNeighbours),
     runState: root.runState,
+    cellLabels: root.cellLabels,
```
Comment (bfollington) [inline packages/cf-harness/console/run-store.ts]: 🔴 Blocking (verified): the family map uses each child's own label index for its handles, but discards every `child.cellLabels` summary here and presents only the root status as the regime for the whole map. In a mixed family (for example root `read`, child `unavailable`, or root `absent` with a child `read`), a child's bare cell is interpreted under the wrong run-level fact—the exact read-vs-unavailable distinction this artifact is meant to preserve. Carry status per run/node, or expose a conservative mixed/partial family summary rather than claiming the root's status for all descendants. Please cover both mixed-status directions.

File: packages/cf-harness/console/server.ts [hunk 4]
```diff
@@ -91,6 +91,9 @@ import {
 } from "../src/sandbox/docker-runsc.ts";
 import type { CreateHarnessPromptLoopOptions } from "../src/prompt-loop.ts";
 import type { HarnessChatSessionStore } from "../src/session-store.ts";
+import { FileSystemHarnessArtifactStore } from "../src/artifacts.ts";
+import type { HarnessRunState } from "../src/run-state.ts";
+import { readSpaceCellLabels } from "../src/space-labels.ts";
```
Comment (bfollington) [inline packages/cf-harness/console/server.ts]: 🔴 Blocking (verified): this static import eagerly evaluates `state-inspector` and `@db/sqlite` in every console process, including zero-ref runs for which `#recordCellLabels()` takes no snapshot. On the exact PR tree, importing `engine.ts` and exercising both no-session and zero-ref exits succeeds without `--allow-ffi`, while importing only `console/server.ts` fails in `@db/sqlite/src/ffi.ts` for missing FFI permission. This also contradicts the nearby `--session-db none` promise that a machine without the native library can run. Move the dynamic import inside `#recordCellLabels()` after the `refs.length === 0` guard (cache the module promise if useful), and add a subprocess import test without FFI permission.

File: packages/cf-harness/console/server.ts [hunk 5]
```diff
@@ -679,6 +694,22 @@ export class ConsoleServer {
...
+    if (
+      envelope.event.kind === "turn_completed" ||
+      envelope.event.kind === "turn_failed" ||
+      envelope.event.kind === "turn_canceled"
+    ) {
+      // The run's cells are settled once its turn is, and the page re-reads
+      // the run on the same event, so the snapshot is taken here rather than
+      // on the read that wants it. A failure to read the space is a snapshot
+      // that says so; a failure to write one leaves the run as it stands.
+      this.#recordCellLabels(envelope.sessionId).catch((error: unknown) => {
```
Comment (bfollington) [inline packages/cf-harness/console/server.ts]: 🔴 Blocking (verified): the snapshot is fire-and-forget, then this same `broadcast()` immediately exposes the terminal event to SSE clients. The frontend refreshes the run/map once on `turn_completed` or `turn_failed`, but no event is emitted when `cell-labels.json` finishes writing. Therefore that refresh can observe `absent` and remain stale until another event or a manual reload. Make snapshot completion happen before the terminal event becomes observable, or emit a dedicated artifact-updated event and refresh on it. An event-controlled test should hold the snapshot promise unresolved and prove clients cannot perform their final read early; no sleeps are needed.

File: packages/cf-harness/src/space-labels.ts [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+  const linked: { key: string; id: string }[] = [];
+  for (const [key, held] of Object.entries(value)) {
+    const link = decodedLinkOf(held as never);
+    if (link?.id === undefined || link.id === null) {
+      continue;
+    }
+    // A link naming another space addresses a store this reader did not open,
+    // and one naming a space against an unknown DID cannot be shown to be
+    // this one. Only a link naming this space, or naming none, is a cell here.
+    if (addressesOpenedSpace(link.space, space)) {
```
Comment (bfollington) [inline packages/cf-harness/src/space-labels.ts]: 🔴 Blocking (verified): direct cross-space refs now fail closed correctly, but the one-hop case only drops the link here. That skip is not represented in `HarnessCellLabels`: the parent record stays readable, contains no entry for the linked path, and is indexed by the console. A negative probe then resolves `/<parent>/foreignResult` to a defined `ConsoleCellLabels` with `entries: []`, so the card says “the space holds no label for this cell” even though the target was never read. This also silently hides a legitimate same-space link whenever an explicit `--space-db` has a non-DID basename, because every explicitly spaced link is unprovable and dropped.

Please preserve per-link/per-path unread coverage in the artifact (or conservatively mark the containing record unavailable), propagate it through `cellLabelsAt()` and the family summary, and cover both a foreign one-hop collision and an unprovable same-space link. The current README statement that a foreign “reference or link” is recorded unread is true only for direct references.

File: packages/cf-harness/console/cell-labels.ts [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+ * written in: `*` stands for every member of a container, so it matches any
+ * concrete segment and any concrete segment matches it.
+ *
+ * Mirrors `cfcLabelPathPrefixMatches` in
+ * `packages/runner/src/cfc/label-view-core.ts`, which is the definition of
+ * this relation and of the re-rooting below it in `rebaseCfcLabelView`. It is
+ * not exported from `@commonfabric/runner/cfc`, so it is restated here; the
+ * two are one rule and a change to that one belongs here too.
+ */
+const pathPrefixMatches = (
```
Comment (bfollington) [inline packages/cf-harness/console/cell-labels.ts]: 🔴 Blocking (verified): the requested wildcard, RFC 6901 escape, and ordinary sibling tests pass, but this local copy is not the full CFC rebase semantics the comment claims it mirrors. Against the PR head, a stored `['secret']` entry narrows to no atoms for the valid `/<entity>/value/secret` reference, while canonical `rebaseCfcLabelView()` returns it because CFC canonicalizes the leading `value` segment. In the other direction, a `shape`-only entry at `['container']` is inherited by `container/public`, while canonical rebase excludes a non-content observation on a strict ancestor. Those are respectively a hidden label and a grafted label.

Please reuse the canonical rebase machinery (or expose the exact projection needed here) instead of restating only `cfcLabelPathPrefixMatches`, and add the leading-`value` and strict-ancestor observation cases. `@commonfabric/runner/cfc` exports `rebaseCfcLabelView`, and the package also exports `./cfc/label-view-core`; the claim that the definition is unavailable is not accurate.

**Review by bfollington — COMMENTED** [review]
cf-review via Codex (GPT-5), on behalf of @bfollington

Scope: PR #6472 re-review after the five blocking fixes · deep-read of the fix commit and immediate label-reader/console/event-stream blast radius.

🔴 Blocking: two verified findings remain and are anchored inline. Direct cross-space records now fail closed, preserve `unreadReason`, and stay out of the console index, but a skipped one-hop link loses the unread fact and is presented as a recorded empty label set. The new segment matcher fixes `*`, JSON Pointer escapes, and ordinary siblings, but still diverges from canonical CFC rebase for the leading `value` segment and observation-class inheritance.

Closed: mixed family statuses report `unavailable` with the split; the console loads the space reader only after the no-ref guard; terminal chat events wait for snapshot settlement, later chat events stay ordered, and snapshot failure does not drop the event.

Checks: focused tests passed (4 files, 136 steps); three negative probes reproduced the remaining one-hop unread loss and both canonical-rebase divergences; fix-commit diff check, touched-file format, and touched-file lint passed. The no-FFI subprocess uses the repository’s isolated-Deno helper and a temporary lock.

Verdict: changes are still required before merge.

File: packages/cf-harness/src/space-labels.ts [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+  linked: { key: string; id: string }[];
+  unreadPaths: HarnessCellLabelUnreadPath[];
+} => {
+  const value = document?.value;
+  if (!isRecord(value)) {
+    return { linked: [], unreadPaths: [] };
+  }
+  const linked: { key: string; id: string }[] = [];
+  const unreadPaths: HarnessCellLabelUnreadPath[] = [];
+  for (const [key, held] of Object.entries(value)) {
```
Comment (bfollington) [inline packages/cf-harness/src/space-labels.ts]: 🔴 Blocking (verified): this only asks `decodedLinkOf()` about direct properties of `document.value`. A link nested inside an object or array is still one hop from this document, but it is neither followed nor added to `unreadPaths`. I seeded `value.nested.theirs` with a foreign-space link; the reader returned only `[{ path: ["theirs"], reason: "cross-space" }]`, and the full `readSpaceCellLabels()` → `cellLabelsAt("/<parent>/nested/theirs")` path returned a defined empty `ConsoleCellLabels`. The card therefore says “the space holds no label for this cell” although the target was never read.

Please walk the document value recursively until each sigil link (without recursing into linked targets), retain the full logical path for both followed entries and unread paths, and cover nested same-space plus cross-space links. The new prefix logic already suppresses exactly that path and its descendants once the artifact records it.

File: packages/cf-harness/console/cell-labels.ts [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+ * One stored entry as a label view of its own, which is the form
+ * {@link rebaseCfcLabelView} narrows. One entry per view rather than the
+ * whole set in one: the canonical rebase merges the labels of entries that
+ * land on one path, and `origin` and the provenance beside it belong to the
+ * entry that carried them, so a merged label could no longer say which of
+ * them was computed and which was declared.
+ *
+ * The atoms are the ones the space stored. The contract widens them to carry
+ * fields no reader has a name for, which the atom types do not describe, and
+ * the rebase reads no atom's interior — only whether a label carries any. So
```
Comment (bfollington) [inline packages/cf-harness/console/cell-labels.ts]: 🟡 Improvement (verified): the cast is operationally safe for the JSON-backed artifact shapes I probed, including the display-normalized `anyOf` wrapper and opaque nested fields, but this justification is not accurate. `rebaseCfcLabelView()` ends in `mergeCfcLabelViews()`, whose `mergeLabel()` normalizes confidentiality clauses and structurally deduplicates atoms, so it does inspect their interiors.

Because this caller discards `rebased.label` and needs only path inclusion/re-rooting, consider supplying a typed non-empty sentinel `IFCLabel` whenever the artifact entry carries any label. That removes the double cast and makes the path-only dependency true. At minimum, update the comment to describe the normalization/deduplication that actually occurs.

**Review by bfollington — COMMENTED** [review]
cf-review via Codex (GPT-5), on behalf of @bfollington

Scope: PR #6472 third pass at `2e864bf528` · deep-read of the two fix areas and their artifact → index → lookup → card path.

🔴 Blocking: one verified finding remains, anchored inline. `unreadPaths` is correct and precisely scoped once a path reaches the artifact, but `linkedCellsOf()` only inspects direct properties of `document.value`. A nested link is still one hop away and is never followed or marked unread; the end-to-end probe resolves that nested target to a defined empty label set, so the card renders it as read-and-unlabelled.

🟡 Improvement: the `IFCLabel` cast is operationally safe for the generated JSON artifact shapes tested, but its comment is inaccurate. Canonical rebase reaches `mergeLabel()`, which normalizes clauses and structurally deduplicates atoms. A typed non-empty sentinel would make the actual path-only dependency explicit and remove the double cast.

Closed: canonical rebase now owns inclusion, leading-`value`, wildcard, and observation-class behavior. Its loop processes each source entry independently and produces at most one rebased entry; the terminal merge is the only cross-entry operation, so singleton rebasing preserves the same inclusion and re-rooted path while avoiding coalescence. A same-path declared/derived probe kept both origins and the derived producer. For recorded unread paths, the path and descendants return `undefined`, siblings remain readable, and the containing cell remains readable while retaining its unread descendants.

Checks: the unmodified focused tests passed (2 files, 86 steps); touched-file format, lint, and focused check passed under Deno 2.9.4. Additional provenance and normalized/opaque-atom probes passed. Two negative nested-link probes failed as expected: the full unread path was absent, and `cellLabelsAt()` returned `{ entries: [] }` instead of `undefined`.

Verdict: changes are still required before merge.

File: packages/cf-harness/src/space-labels.ts (hunk 8 above)
Comment (bfollington) [reply packages/cf-harness/src/space-labels.ts]: 🔴 Still blocking (verified): the recursive shape enumeration fixes the concrete nested object/array case from this thread, but the finite bounds silently recreate the same read-versus-unread failure. `linksWithPaths()` returns only `found`, so `linkedCellsOf()` cannot tell a complete walk from one stopped by `maxDepth: 64` or `maxNodes: 100_000` and returns no `unreadPaths` for the skipped subtree. A foreign link at depth 65 produced 0 bounded results versus 1 unbounded result; so did a foreign link in the final slot of a 100,000-item array (the root plus preceding items consume the budget). Both are valid finite JSON values. The artifact therefore records the parent as read with no fact at that path, and `cellLabelsAt()` can again render the target as read-and-unlabelled. The README claim that links are found however deep is also unsupported.

Please return completion/truncation from the walk and make the label reader turn any truncation into an unread fact for the skipped subtree, or conservatively mark the whole record unread. Add reader-level tests for both depth and node-budget exhaustion.

File: packages/cf-harness/console/cell-labels.ts (hunk 9 above)
Comment (bfollington) [reply packages/cf-harness/console/cell-labels.ts]: ✅ Closed (verified): `ENTRY_PRESENT` is a typed, non-empty `IFCLabel`; clone/merge and structural dedup leave its sole integrity atom non-empty. The caller consumes only inclusion and the rebased path, then maps display atoms, origin, and provenance from the original artifact entry through `entryOf()`. No stored atom crosses the rebase, and an originally empty entry is still dropped.

**Review by bfollington — COMMENTED** [review]
cf-review via Codex (GPT-5), on behalf of @bfollington

Scope: PR #6472 fourth pass at `c23aba2192` · deep-read of the one-hop walker, both state-inspector call sites, the label-reader integration, and the sentinel rebase.

🔴 Blocking: one verified finding remains, continued on the existing nested-link thread. The walker is generic over recognized link shape and covers the value root, object properties, and array indices, but its finite depth and node bounds silently return partial results. The label reader receives no truncation fact, so a link skipped at depth 65 or after the 100,000-node budget can still render as read-and-unlabelled. The live README claim that links are found however deep is false for the same reason.

Closed: the typed non-empty sentinel is safe through canonical clone/merge/dedup, no stored atom crosses the rebase, and empty stored entries remain empty. `detail.ts` preserves the previous depth-10 behavior at both call sites, including traversal order, root/array handling, stopping at links, and final joined display paths.

Assessment: `collectLinks()` and `countLinks()` are legitimately separate from the new path-aware walker. They deliberately stop at stream/entity-ref leaves, traverse plain objects, use different depth defaults, and the counter avoids result/path allocation; folding them would require semantic policy switches or change behavior.

Checks: 6 focused test files passed (137 steps); touched-file format and lint passed. Direct probes confirmed modern/legacy root and nested shape coverage, old-vs-new detail equivalence, sentinel behavior by source, and the two silent truncation cases.

Verdict: changes are still required before merge.

File: packages/cf-harness/src/space-labels.ts (hunk 8 above)
Comment (bfollington) [reply packages/cf-harness/src/space-labels.ts]: 🔴 Still blocking (verified): the walk and artifact are honest now, but the console presentation still turns the partial record into the precise negative claim this thread is about. `cellLabelsAt()` correctly carries `truncationReason: "node-budget-exhausted"` through every narrowed path, but `cellLabelView()` drops both `truncationReason` and `unreadPaths`, sets `recorded: true`, and the empty-atom branch in `cell-chip.ts` renders **“the space holds no label for this cell.”** `consoleCellLabelsSummary()` likewise ignores both partial facts, so an empty truncated record produces `status: "read", cellsRead: 1, cellsLabelled: 0`, and the family header says **“cell labels read · 0 of 1 carry one.”**

The same UI failure remains for a depth stop when the rendered reference is an ancestor of the declined path: the declined path itself correctly returns `undefined`, but the root/ancestor returns a recorded empty label object carrying `unreadPaths`; the chip drops that field and makes the same unlabelled claim.

I reproduced both cases against `054c4e37de`: the budget case retained `truncationReason` at `/any/path`, and the depth case retained `unreadPaths: [["deep", "down"]]` at the root, but both reduced to `{ recorded: true, confidentiality: [], integrity: [], paths: [] }` in `cellLabelView()`, while both summaries reported 0 of 1 labelled. Please carry an explicit partial/unknown fact into `ConsoleCellLabelView` and `ConsoleCellLabelsSummary` (including the family fold), render it on the chip/header, and reserve “the space holds no label” for records with neither `unreadPaths` nor `truncationReason`. Add chip and family/header tests for empty depth-partial and budget-partial records.

**Review by bfollington — COMMENTED** [review]
cf-review via Codex (GPT-5), on behalf of @bfollington

Scope: PR #6472 fifth pass at `054c4e37de` · deep-read of the bounded walker, artifact contract, reader, console index/narrowing/card/summary/family paths, and both `detail.ts` call sites.

🔴 Blocking: one verified finding remains, continued on the existing truncation thread. The walker and artifact now distinguish exact depth refusals from record-level budget exhaustion, but the console drops both partial facts at its final presentation boundary. A budget-truncated empty record reaches every `cellLabelsAt()` path with `truncationReason`, yet `cellLabelView()` reduces it to recorded-and-empty and the card says “the space holds no label for this cell”; the summary/family header says “cell labels read · 0 of 1 carry one.” A depth-partial root or ancestor does the same because `unreadPaths` is also absent from the chip view and summary, although a reference at or below the declined path correctly returns no labels.

Closed below that boundary: `linksWithPaths()` returns links plus exact depth stops and the weaker budget fact; budget is checked first; the reader records `below-read-depth`, records and warns on `node-budget-exhausted`, and the injectable finite bounds cannot silently narrow the result. `narrow()` propagates record-level truncation to every path, not only the root. `detail.ts` remains behaviorally identical: its only operational change is selecting `.links` at the two existing depth-10, unbounded-node call sites, and the link enumeration itself is unchanged.

Checks: 8 focused test files passed (195 steps) under Deno 2.9.4; touched-file format, lint, and focused check passed. Direct console probes reproduced both remaining false-negative presentations. All production callers of `linksWithPaths()` were inspected.

Verdict: changes are still required before merge.

File: packages/cf-harness/console/run-store.ts [hunk 10]
```diff
@@ -334,9 +383,71 @@ const readConsoleRunFamily = async (
...
+ * The cells themselves are merged by address, so a cell a parent and its
+ * child both hold is one cell of the family rather than two.
+ */
+const familyCellLabels = (
+  members: readonly FamilyLabelReading[],
+): ConsoleCellLabelsSummary => {
+  const byAddress = new Map<string, HarnessCellLabelRecord>();
+  for (const member of members) {
+    for (const [address, record] of member.labels.byAddress) {
+      byAddress.set(address, record);
```
Comment (bfollington) [inline packages/cf-harness/console/run-store.ts]: _cf-review via Codex (GPT-5), on behalf of @bfollington_

🔴 **Still blocking (verified at `68348123b7`): duplicate family addresses can still erase the partial fact.** This map keeps only one whole `HarnessCellLabelRecord` per address, but `set()` replaces the earlier member's record rather than combining the readings. I reproduced a root whose `/of:…/same` record has `truncationReason: "node-budget-exhausted"` and a child that later records the same address whole: the family summary returns `cellsRead: 1, cellsPartial: 0`. The root's flow card still correctly carries its partial row, while the family header now says `cell labels read · …` and counts that cell as fully read.

The graph has the same first/last-reading problem in the other direction: at `console/graph.ts:405-407`, the first non-empty labels object always wins. I reproduced a root delegating in order to two children that touch the same address—the first child's reading is whole and the later child's is truncated—and the coalesced graph node returns `truncationReason: undefined`. Flow's per-node copying is otherwise sound, the new reading row survives when atoms are present, and a genuinely complete empty record still takes the correct bare branch.

Please fold duplicate-address readings conservatively instead of choosing one: for the summary, an address is partial if any member's reading is partial and labelled if any member found an entry; for the family graph, merge the label facts while preserving any `unreadPaths`/`truncationReason`. Add parent/child same-address tests in both directions. Until then there are still header and graph paths by which a cell one reader did not fully read is represented as fully read.

**Review by bfollington — COMMENTED** [review]
_cf-review via Codex (GPT-5), on behalf of @bfollington_

Scope: PR #6472 seventh pass at `e8bc52f54f` · deep-read of the shared fold, both family consumers, the presentation predicate, and the artifact/index boundaries that feed them.

No blocking findings or improvements remain in this pass.

Closed, specifically:

- **No remaining partial-to-full route found.** Every defined reading that carries `unreadPaths` or `truncationReason` reaches `foldCellLabels`; both facts survive through set union/disjunction into the coalesced graph node and the family summary, so the card remains partial and `cellsPartial` remains nonzero regardless of which family member arrives first. A sighting with no label record is still treated as no reading, not as a competing complete reading; it cannot itself produce the positive empty-space claim.
- **The fold is genuinely order-independent beyond pairs.** Entries and unread paths are canonically ordered set unions, truncation is a fixed-order disjunction, and cell-level atoms/derived/producers are recomputed from the folded entries. I probed four readings through all 24 permutations with left, right, and balanced association; every result was identical.
- **Entry identity neither drops nor invents a displayed entry.** Its JSON tuple is injective over every field `ConsoleCellLabelEntry` states: path, sorted confidentiality and integrity atoms, origin, observation class, and producer. Reordered atoms collapse as the same entry; same-path entries differing in origin/observation/provenance remain separate. Delimiter-like path and atom strings do not collide because the key is structured JSON rather than concatenation.
- **The sentence has one runtime producer.** The sole rendered literal `the space holds no label for this cell` is the branch guarded by `spaceHoldsNoLabel`. The predicate requires a recorded, non-partial reading with both atom dimensions empty, and its positive case plus each widening term has a dedicated negative test. Other exact-text occurrences are explanatory comments only.

Checks: 4 focused test files passed (158 steps) under Deno 2.9.4; focused `deno check`, touched-file `deno fmt --check`, touched-file `deno lint`, and fix-commit `git diff --check` passed. The additional algebra/identity probe covered four-member permutation and reassociation cases.

Verdict: **closed; no changes requested.**
